### PR TITLE
ADBDEV-4436: Add PXF integration tests for custom features

### PR DESCRIPTION
**Pull request by RomaZe**

### CODE REVIEW

File: automation/arenadata/Dockerfile [hunk 1]
```diff
@@ -36,6 +36,10 @@ RUN sudo -H -u gpadmin bash -c "pip install --user paramiko"
 RUN localedef -c -i ru_RU -f CP1251 ru_RU.CP1251
 RUN cp ${PXF_HOME}/templates/*-site.xml ${PXF_HOME}/servers/default/
 
+# Download jar files
+RUN curl -fSL https://repo.maven.apache.org/maven2/io/arenadata/security/encryption/1.0.0/encryption-1.0.0-exec.jar -o /usr/local/greenplum-db-devel/pxf/lib/encryption-1.0.0-exec.jar
```
Comment (xardazzzzzz) [inline automation/arenadata/Dockerfile]: I think we should download from our internal repo

File: automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        String queryDropTable = "DROP TABLE " + table.getSchema() + "." + table.getName();
+        runQuery(queryDropTable, true, false);
+    }
+
+    @Override
+    public boolean checkTableExists(Table table) throws Exception {
+        DatabaseMetaData metaData = dbConnection.getMetaData();
+        String tableName = "";
+        if (table != null) {
+            tableName = table.getName().toUpperCase();
```
Comment (xardazzzzzz) [inline automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java:78]: Why do we need to put it to uppercase?

File: automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        }
+        ResultSet res = metaData.getTables(null, null, "%" + tableName, new String[]{"TABLE", "FOREIGN TABLE"});
+        return res.next();
+    }
+
+    public int getValueFromQuery(String query) throws Exception {
+        ReportUtils.report(report, getClass(), "Get value - query: " + query);
+        ResultSet res = stmt.executeQuery(query);
+        res.next();
```
Comment (xardazzzzzz) [inline automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java]: We should check result of res.next()

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    private String pxfHome;
+    private String pxfJdbcSiteConfFile;
+    private String pxfJdbcSiteConfTemplate;
+    private Table gpdbEncryptionSourceTable;
+
+    @Override
+    public void beforeClass() throws Exception {
+        pxfHome = cluster.getPxfHome();
+        pxfJdbcSiteConfFile = pxfHome + "/servers/" + PXF_ENCRYPTION_SERVER_PROFILE + "/jdbc-site.xml";
+        pxfJdbcSiteConfTemplate = pxfHome + "/templates/encryption/jdbc-site.xml";
```
Comment (xardazzzzzz) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java]: Looks like more candidates for constants

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+    private static final String NAMED_QUERY_SERVER_PROFILE = "named";
+    private static final String LONG_YEAR_DATA_FILE_NAME = "long_year.txt";
+    private Table gpdbLongYearSourceTable, gpdbLongYearTargetTable, gpdbLongYearTargetLegacyTable;
+    private Table gpdbBoolDateTypeSourceTable;
+
+    @Override
+    protected void beforeClass() throws Exception {
+        String pxfHome = cluster.getPxfHome();
+        String queryFile = pxfHome + "/templates/named-query/named_query.sql";
```
Comment (xardazzzzzz) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java]: Extract to constants

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    private String pxfHome;
+    private String pxfJdbcSiteConfFile;
+    private String pxfJdbcSiteConfTemplate;
+    private Oracle oracle;
+
+    @Override
+    public void beforeClass() throws Exception {
+        pxfHome = cluster.getPxfHome();
+        pxfJdbcSiteConfFile = pxfHome + "/servers/" + PXF_ORACLE_SERVER_PROFILE + "/jdbc-site.xml";
+        pxfJdbcSiteConfTemplate = pxfHome + "/templates/oracle/jdbc-site.xml";
```
Comment (xardazzzzzz) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java]: Here too

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    private String pxfHome;
+    private String pxfJdbcSiteConfFile;
+    private String pxfLogFile;
+    private Table gpdbPredicateInSourceTable, oraclePredicateInSourceTable;
+
+    @Override
+    protected void beforeClass() throws Exception {
+        pxfHome = cluster.getPxfHome();
+        pxfJdbcSiteConfFile = pxfHome + "/servers/" + PXF_ORACLE_SERVER_PROFILE + "/jdbc-site.xml";
+        String pxfJdbcSiteConfTemplate = pxfHome + "/templates/oracle/jdbc-site.xml";
```
Comment (xardazzzzzz) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: Seems like it's a global constant that can be extracted to separate class/interface

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            "birth_date_dad  timestamp"};
+
+    private static final String[] NAMED_QUERY_TABLE_FIELDS = new String[]{
+            "id  int",
+            "name text"};
+
+    private static final String NAMED_QUERY_SERVER_PROFILE = "named";
+    private static final String LONG_YEAR_DATA_FILE_NAME = "long_year.txt";
+    private Table gpdbLongYearSourceTable, gpdbLongYearTargetTable, gpdbLongYearTargetLegacyTable;
+    private Table gpdbBoolDateTypeSourceTable;
```
Comment (SidorovGreg) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java:41]: Why this one on another line?

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+    @Test(groups = {"arenadata"}, description = "Check pushdown predicate 'IN' for Postres")
+    public void testPredicateInPostgres() throws Exception {
+        // Clean gpdb master log and pxf log before run test
+        cluster.runCommand("> " + GET_LATEST_MASTER_LOG_COMMAND);
+        cluster.runCommandOnNodes(Collections.singletonList(pxfNode), "> " + pxfLogFile);
+        runTincTest("pxf.arenadata.predicate-in.postgres.runTest");
+        cluster.runCommand("grep -e 'SELECT id, descr FROM " + SOURCE_TABLE_NAME + " WHERE id IN (2,3)' " + GET_LATEST_MASTER_LOG_COMMAND + " | wc -l");
+        String result = cluster.getLastCmdResult();
+        assertEquals("1", result.split("\r\n")[1].trim());
```
Comment (SidorovGreg) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: We cannot be sure that there is 2 element. There will be a null pointer which is not obvious

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    @Test(groups = {"arenadata"}, dependsOnMethods={"testPredicateInPostgres"}, description = "Check pushdown predicate 'IN' logging")
+    public void testPredicateInLogging() throws Exception {
+        cluster.copyFromRemoteMachine(pxfNode.getUserName(), pxfNode.getPassword(), pxfNode.getHost(), pxfLogFile, "/tmp/");
+        cluster.runCommand(POSTGRES_SEGMENT_LOG_GREP_COMMAND);
+        String result = cluster.getLastCmdResult();
+        assertEquals("1", result.split("\r\n")[1].trim());
```
Comment (SidorovGreg) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: We can't be sure that there is 2 element

File: automation/arenadata/Dockerfile (hunk 1 above)
Comment (RomaZe) [reply automation/arenadata/Dockerfile]: Why should we use our internal repo in this project?

File: automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java (hunk 2 above)
Comment (RomaZe) [reply automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java:78]: It is not possible to find the table by name in lower case. Jdbc driver uses `OracleDatabaseMetaData#getTables` method to get the table name. And tables name are saved in upper case format.

File: automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java (hunk 3 above)
Comment (RomaZe) [reply automation/src/main/java/org/greenplum/pxf/automation/components/oracle/Oracle.java]: Added condition to check res.next()

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java (hunk 4 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java]: `pxfJdbcSiteConfFile` and `pxfJdbcSiteConfTemplate` depends on `pxfHome` which is not static field.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java (hunk 5 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java]: `queryFile` depends on pxfHome which is not static field.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java (hunk 6 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java]: `pxfJdbcSiteConfFile` and `pxfJdbcSiteConfTemplate` depends on `pxfHome` which is not static field.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java (hunk 7 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: I am not quite sure that we need to create an additional class for one global constant which has already been defined as a `pxfHome` field in the `PhdCluster.class`. The others fields `pxfJdbcSiteConfFile` and  `pxfJdbcSiteConfTemplate` are different for different classes.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java (hunk 10 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: Add the condition to check that the result has at least 2 elements

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java (hunk 9 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: Add the condition to check that the result has at least 2 elements

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java (hunk 8 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java:41]: Moved each variable to a separate line

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java (hunk 5 above)
Comment (xardazzzzzz) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java]: I'm talking about static part that should be moved to constant

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java (hunk 4 above)
Comment (xardazzzzzz) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java]: I'm talking about static part that should be moved to constant

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java (hunk 6 above)
Comment (xardazzzzzz) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java]: I'm talking about static part that should be moved to constant

Comment (xardazzzzzz) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java]: Also as these static parts are used in multiple places consider extracting them to separate class

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java (hunk 7 above)
Comment (xardazzzzzz) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: `"/servers/" + PXF_ORACLE_SERVER_PROFILE + "/jdbc-site.xml";` is a common part for multiple classes

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java (hunk 4 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcEncryptionTest.java]: Added constant

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java (hunk 5 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/JdbcFeaturesTest.java]: Added constant

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java (hunk 6 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/OracleParallelSessionTest.java]: Added constant for `/templates/oracle/jdbc-site.xml`

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java (hunk 7 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java]: `PXF_ORACLE_SERVER_PROFILE` variable has a different value in the different classes.

**Review by xardazzzzzz — APPROVED** [review]

**Pull request merged**